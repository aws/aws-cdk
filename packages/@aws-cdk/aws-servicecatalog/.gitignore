*.js
*.js.map
*.d.ts
tsconfig.json
node_modules
product-stack-snapshots
*.generated.ts
dist
.jsii

.LAST_BUILD
.nyc_output
coverage
nyc.config.js
.LAST_PACKAGE
*.snk
!.eslintrc.js
!jest.config.js

junit.xml
!**/*.snapshot/**/asset.*/*.js
!**/*.snapshot/**/asset.*/*.d.ts

!**/*.snapshot/**/asset.*/**
