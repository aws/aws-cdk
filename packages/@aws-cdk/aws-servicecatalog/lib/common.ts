/**
 * The language code.
 */
export enum AcceptLanguage {
  /**
   * English
   */
  EN = 'en',

  /**
   * Japanese
   */
  JP = 'jp',

  /**
   * Chinese
   */
  ZH = 'zh'
}
