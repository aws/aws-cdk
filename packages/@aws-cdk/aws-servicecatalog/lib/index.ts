export * from './common';
export * from './cloudformation-template';
export * from './portfolio';
export * from './product';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
