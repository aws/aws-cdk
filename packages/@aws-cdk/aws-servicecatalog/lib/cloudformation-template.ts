import * as s3_assets from '@aws-cdk/aws-s3-assets';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Represents the Product Provisioning Artifact Template.
 */
export abstract class CloudFormationTemplate {
  /**
   * Template from URL
   * @param url The url that points to the provisioning artifacts template
   */
  public static fromUrl(url: string): CloudFormationTemplate {
    return new CloudFormationUrlTemplate(url);
  }

  /**
   * Loads the provisioning artifacts template from a local disk path.
   *
   * @param path A file containing the provisioning artifacts
   */
  public static fromAsset(path: string, options?: s3_assets.AssetOptions): CloudFormationTemplate {
    return new CloudFormationAssetTemplate(path, options);
  }

  /**
   * Called when the product is initialized to allow this object to bind
   * to the stack, add resources and have fun.
   *
   * @param scope The binding scope. Don't be smart about trying to down-cast or
   * assume it's initialized. You may just use it as a construct scope.
   */
  public abstract bind(scope: Construct): CloudFormationTemplateConfig;
}

/**
 * Result of binding `Template` into a `Product`.
 */
export interface CloudFormationTemplateConfig {
  /**
    * The http url of the template in S3.
    */
  readonly httpUrl: string;
}

/**
 * Template code from a Url.
 */
class CloudFormationUrlTemplate extends CloudFormationTemplate {
  constructor(private readonly url: string) {
    super();
  }

  public bind(_scope: Construct): CloudFormationTemplateConfig {
    return {
      httpUrl: this.url,
    };
  }
}

/**
 * Template from a local file.
 */
class CloudFormationAssetTemplate extends CloudFormationTemplate {
  private asset?: s3_assets.Asset;

  /**
   * @param path The path to the asset file.
   */
  constructor(public readonly path: string, private readonly options: s3_assets.AssetOptions = { }) {
    super();
  }

  public bind(scope: Construct): CloudFormationTemplateConfig {
    // If the same AssetCode is used multiple times, retain only the first instantiation.
    if (!this.asset) {
      this.asset = new s3_assets.Asset(scope, 'Template', {
        path: this.path,
        ...this.options,
      });
    }

    return {
      httpUrl: this.asset.httpUrl,
    };
  }
}
