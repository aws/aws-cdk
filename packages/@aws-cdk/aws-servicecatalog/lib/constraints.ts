import { MessageLanguage } from './common';

/**
 * Properties for governance mechanisms and constraints.
 */
export interface CommonConstraintOptions {
  /**
   * The language code.
   * Configures the language for error messages from service catalog.
   *
   * @default - English
   */
  readonly messageLanguage?: MessageLanguage;

  /**
   * The description of the constraint.
   *
   * @default - No description provided
   */
  readonly description?: string;
}

/**
 * Properties for ResourceUpdateConstraint.
 */
export interface TagUpdateConstraintOptions extends CommonConstraintOptions {
  /**
   * Toggle for if users should be allowed to change/update tags on provisioned products.
   * @default true
   */
  readonly allow?: boolean;
}