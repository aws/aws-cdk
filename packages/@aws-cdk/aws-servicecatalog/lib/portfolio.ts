import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { AcceptLanguage } from './common';
import { hashValues } from './private/util';
import { InputValidator } from './private/validation';
import { CfnPortfolio, CfnPortfolioPrincipalAssociation, CfnPortfolioShare } from './servicecatalog.generated';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from 'constructs';

/**
 * Options for portfolio share.
 */
export interface PortfolioShareOptions {
  /**
   * Whether to share tagOptions as a part of the portfolio share
   * @default - share not specified
   */
  readonly shareTagOptions?: boolean;

  /**
   * The accept language of the share
   * @default - accept language not specified
   */
  readonly acceptLanguage?: AcceptLanguage;
}

/**
 * A Service Catalog portfolio.
 */
export interface IPortfolio extends cdk.IResource {
  /**
   * The ARN of the portfolio.
   * @attribute
   */
  readonly portfolioArn: string;

  /**
   * The ID of the portfolio.
   * @attribute
   */
  readonly portfolioId: string;

  /**
   * Associate portfolio with an IAM Role.
   * @param role an IAM role
   */
  giveAccessToRole(role: iam.IRole): void;

  /**
   * Associate portfolio with an IAM User.
   * @param user an IAM user
   */
  giveAccessToUser(user: iam.IUser): void;

  /**
   * Associate portfolio with an IAM Group.
   * @param group an IAM Group
   */
  giveAccessToGroup(group: iam.IGroup): void;

  /**
   * Initiate a portfolio share with another account.
   * @param accountId AWS account to share portfolio with
   * @param options Options for the initiate share
   */
  shareWithAccount(accountId: string, options?: PortfolioShareOptions): void;
}

abstract class PortfolioBase extends cdk.Resource implements IPortfolio {
  public abstract readonly portfolioArn: string;
  public abstract readonly portfolioId: string;
  private readonly associatedPrincipals: Set<string> = new Set();

  public giveAccessToRole(role: iam.IRole): void {
    this.associatePrincipal(role.roleArn, role.node.addr);
  }

  public giveAccessToUser(user: iam.IUser): void {
    this.associatePrincipal(user.userArn, user.node.addr);
  }

  public giveAccessToGroup(group: iam.IGroup): void {
    this.associatePrincipal(group.groupArn, group.node.addr);
  }

  public shareWithAccount(accountId: string, options: PortfolioShareOptions = {}): void {
    const hashId = this.generateUniqueHash(accountId);
    new CfnPortfolioShare(this, `PortfolioShare${hashId}`, {
      portfolioId: this.portfolioId,
      accountId: accountId,
      shareTagOptions: options.shareTagOptions,
      acceptLanguage: options.acceptLanguage,
    });
  }

  /**
   * Associate a principal with the portfolio.
   * If the principal is already associated, it will skip.
   */
  private associatePrincipal(principalArn: string, principalId: string): void {
    if (!this.associatedPrincipals.has(principalArn)) {
      const hashId = this.generateUniqueHash(principalId);
      new CfnPortfolioPrincipalAssociation(this, `PortolioPrincipalAssociation${hashId}`, {
        portfolioId: this.portfolioId,
        principalArn: principalArn,
        principalType: 'IAM',
      });
      this.associatedPrincipals.add(principalArn);
    }
  }

  /**
   * Create a unique id based off the L1 CfnPortfolio or the arn of an imported portfolio.
   */
  protected abstract generateUniqueHash(value: string): string;
}

/**
 * Properties for a Portfolio.
 */
export interface PortfolioProps {
  /**
   * The name of the portfolio.
   */
  readonly displayName: string;

  /**
   * The provider name.
   */
  readonly providerName: string;

  /**
   * The accept language.
   * @default - No accept language provided
   */
  readonly acceptLanguage?: AcceptLanguage;

  /**
   * Description for portfolio.
   * @default - No description provided
   */
  readonly description?: string;
}

/**
 * A Service Catalog portfolio.
 */
export class Portfolio extends PortfolioBase {
  /**
   * Creates a Portfolio construct that represents an external portfolio.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param portfolioArn the Amazon Resource Name of the existing portfolio.
   */
  public static fromPortfolioArn(scope: Construct, id: string, portfolioArn: string): IPortfolio {
    const arn = cdk.Stack.of(scope).parseArn(portfolioArn);
    const portfolioId = arn.resourceName;

    if (!portfolioId) {
      throw new Error('Missing required Portfolio ID from Portfolio ARN: ' + portfolioArn);
    }

    class Import extends PortfolioBase {
      public readonly portfolioArn = portfolioArn;
      public readonly portfolioId = portfolioId!;

      protected generateUniqueHash(value: string): string {
        return hashValues(this.portfolioArn, value);
      }
    }

    return new Import(scope, id, {
      environmentFromArn: portfolioArn,
    });
  }

  public readonly portfolioArn: string;
  public readonly portfolioId: string;
  private readonly portfolio: CfnPortfolio;

  constructor(scope: Construct, id: string, props: PortfolioProps) {
    super(scope, id);

    this.validatePortfolioProps(props);

    this.portfolio = new CfnPortfolio(this, 'Resource', {
      displayName: props.displayName,
      providerName: props.providerName,
      description: props.description,
      acceptLanguage: props.acceptLanguage,
    });
    this.portfolioId = this.portfolio.ref;
    this.portfolioArn = cdk.Stack.of(this).formatArn({
      service: 'servicecatalog',
      resource: 'portfolio',
      resourceName: this.portfolioId,
    });
  }

  protected generateUniqueHash(value: string): string {
    return hashValues(cdk.Names.nodeUniqueId(this.portfolio.node), value);
  }

  private validatePortfolioProps(props: PortfolioProps) {
    InputValidator.validateLength(this.node.path, 'portfolio display name', 1, 100, props.displayName);
    InputValidator.validateLength(this.node.path, 'portfolio provider name', 1, 50, props.providerName);
    InputValidator.validateLength(this.node.path, 'portfolio description', 0, 2000, props.description);
  }
}
