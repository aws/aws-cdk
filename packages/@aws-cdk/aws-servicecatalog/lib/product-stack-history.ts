import * as fs from 'fs';
import * as path from 'path';
import { Names } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CloudFormationTemplate } from './cloudformation-template';
import { DEFAULT_PRODUCT_STACK_SNAPSHOT_DIRECTORY } from './common';
import { CloudFormationProductVersion } from './product';
import { ProductStack } from './product-stack';

/**
 * Properties for a ProductStackHistory.
 */
export interface ProductStackHistoryProps {
  /**
   * The ProductStack whose history will be retained as a snapshot
   */
  readonly productStack: ProductStack;

  /**
   * The current version name of the ProductStack.
   */
  readonly currentVersionName: string;

  /**
   * If this is set to true, the ProductStack will not be overwritten if a snapshot is found for the currentVersionName.
   */
  readonly currentVersionLocked: boolean

  /**
   * The description of the product version
   * @default - No description provided
   */
  readonly description?: string;

  /**
   * Whether the specified product template will be validated by CloudFormation.
   * If turned off, an invalid template configuration can be stored.
   * @default true
   */
  readonly validateTemplate?: boolean;

  /**
   * The directory where template snapshots will be stored
   * @default 'product-stack-snapshots'
   */
  readonly directory?: string
}

/**
 * A Construct that contains a Service Catalog product stack with its previous deployments maintained.
 */
export class ProductStackHistory extends Construct {
  private readonly props: ProductStackHistoryProps
  constructor(scope: Construct, id: string, props: ProductStackHistoryProps) {
    super(scope, id);
    props.productStack._setParentProductStackHistory(this);
    this.props = props;
  }

  /**
   * Retains product stack template as a snapshot when deployed and
   * retrieves a CloudFormationProductVersion for the current product version.
   */
  public currentVersion() : CloudFormationProductVersion {
    return {
      cloudFormationTemplate: CloudFormationTemplate.fromProductStack(this.props.productStack),
      productVersionName: this.props.currentVersionName,
      description: this.props.description,
    };
  }

  /**
   * Retrieves a CloudFormationProductVersion from a previously deployed productVersionName.
   */
  public versionFromSnapshot(productVersionName: string) : CloudFormationProductVersion {
    const productStackSnapshotDirectory = this.props.directory || DEFAULT_PRODUCT_STACK_SNAPSHOT_DIRECTORY;
    const templateFileKey = `${Names.uniqueId(this)}.${this.props.productStack.artifactId}.${productVersionName}.product.template.json`;
    const templateFilePath = path.join(productStackSnapshotDirectory, templateFileKey);
    if (!fs.existsSync(templateFilePath)) {
      throw new Error(`Template ${templateFileKey} cannot be found in ${productStackSnapshotDirectory}`);
    }
    return {
      cloudFormationTemplate: CloudFormationTemplate.fromAsset(templateFilePath),
      productVersionName: productVersionName,
      description: this.props.description,
    };
  }

  /**
   * Writes current template generated from Product Stack to a snapshot directory.
   *
   * @internal
   */
  public _writeTemplateToSnapshot(cfn: string) {
    const productStackSnapshotDirectory = this.props.directory || DEFAULT_PRODUCT_STACK_SNAPSHOT_DIRECTORY;
    if (!fs.existsSync(productStackSnapshotDirectory)) {
      fs.mkdirSync(productStackSnapshotDirectory, { recursive: true });
    }
    const templateFileKey = `${Names.uniqueId(this)}.${this.props.productStack.artifactId}.${this.props.currentVersionName}.product.template.json`;
    const templateFilePath = path.join(productStackSnapshotDirectory, templateFileKey);
    if (fs.existsSync(templateFilePath)) {
      const previousCfn = fs.readFileSync(templateFilePath).toString();
      if (previousCfn !== cfn && this.props.currentVersionLocked) {
        throw new Error(`Template has changed for ProductStack Version ${this.props.currentVersionName}.
        ${this.props.currentVersionName} already exist in ${productStackSnapshotDirectory}.
        Since locked has been set to ${this.props.currentVersionLocked},
        Either update the currentVersionName to deploy a new version or deploy the existing ProductStack snapshot.
        If ${this.props.currentVersionName} was unintentionally synthesized and not deployed, 
        delete the corresponding version from ${productStackSnapshotDirectory} and redeploy.`);
      }
    }
    fs.writeFileSync(templateFilePath, cfn);
  }
}
