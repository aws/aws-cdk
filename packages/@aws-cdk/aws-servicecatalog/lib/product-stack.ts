import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import { Construct } from 'constructs';
import { ProductStackSynthesizer } from './private/product-stack-synthesizer';
import { ProductStackHistory } from './product-stack-history';

/**
 * A Service Catalog product stack, which is similar in form to a Cloudformation nested stack.
 * You can add the resources to this stack that you want to define for your service catalog product.
 *
 * This stack will not be treated as an independent deployment
 * artifact (won't be listed in "cdk list" or deployable through "cdk deploy"),
 * but rather only synthesized as a template and uploaded as an asset to S3.
 *
 */
export class ProductStack extends cdk.Stack {
  public readonly templateFile: string;
  private _parentProductStackHistory?: ProductStackHistory;
  private _templateUrl?: string;
  private _parentStack: cdk.Stack;

  constructor(scope: Construct, id: string) {
    super(scope, id, {
      synthesizer: new ProductStackSynthesizer(),
    });

    this._parentStack = findParentStack(scope);

    // this is the file name of the synthesized template file within the cloud assembly
    this.templateFile = `${cdk.Names.uniqueId(this)}.product.template.json`;
  }

  /**
   * Set the parent product stack history
   *
   * @internal
   */
  public _setParentProductStackHistory(parentProductStackHistory: ProductStackHistory) {
    return this._parentProductStackHistory = parentProductStackHistory;
  }

  /**
   * Fetch the template URL.
   *
   * @internal
   */
  public _getTemplateUrl(): string {
    return cdk.Lazy.uncachedString({ produce: () => this._templateUrl });
  }

  /**
   * Synthesize the product stack template, overrides the `super` class method.
   *
   * Defines an asset at the parent stack which represents the template of this
   * product stack.
   *
   * @internal
   */
  public _synthesizeTemplate(session: cdk.ISynthesisSession): void {
    const cfn = JSON.stringify(this._toCloudFormation(), undefined, 2);
    const templateHash = crypto.createHash('sha256').update(cfn).digest('hex');

    this._templateUrl = this._parentStack.synthesizer.addFileAsset({
      packaging: cdk.FileAssetPackaging.FILE,
      sourceHash: templateHash,
      fileName: this.templateFile,
    }).httpUrl;

    if (this._parentProductStackHistory) {
      this._parentProductStackHistory._writeTemplateToSnapshot(cfn);
    }

    fs.writeFileSync(path.join(session.assembly.outdir, this.templateFile), cfn);
  }
}

/**
 * Validates the scope for a product stack, which must be defined within the scope of another `Stack`.
 */
function findParentStack(scope: Construct): cdk.Stack {
  try {
    const parentStack = cdk.Stack.of(scope);
    return parentStack as cdk.Stack;
  } catch (e) {
    throw new Error('Product stacks must be defined within scope of another non-product stack');
  }
}
