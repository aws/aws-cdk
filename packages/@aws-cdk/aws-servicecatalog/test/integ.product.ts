import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as servicecatalog from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-servicecatalog-product');

new servicecatalog.CloudFormationProduct(stack, 'TestProduct', {
  productName: 'testProduct',
  owner: 'testOwner',
  productVersions: [
    {
      validateTemplate: false,
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromUrl(
        'https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template'),
    },
    {
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromAsset(path.join(__dirname, 'development-environment.template.json')),
    },
  ],
});

app.synth();
