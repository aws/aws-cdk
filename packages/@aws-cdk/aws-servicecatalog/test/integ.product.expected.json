{
  "Resources": {
    "TestProduct7606930B": {
      "Type": "AWS::ServiceCatalog::CloudFormationProduct",
      "Properties": {
        "Name": "testProduct",
        "Owner": "testOwner",
        "ProvisioningArtifactParameters": [
          {
            "DisableTemplateValidation": true,
            "Info": {
              "LoadTemplateFromURL": "https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template"
            }
          },
          {
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Join": [
                  "",
                  [
                    "https://s3.",
                    {
                      "Ref": "AWS::Region"
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix"
                    },
                    "/",
                    {
                      "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3BucketCF629374"
                    },
                    "/",
                    {
                      "Fn::Select": [
                        0,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE"
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Fn::Select": [
                        1,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                ]
              }
            }
          }
        ]
      }
    }
  },
  "Parameters": {
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3BucketCF629374": {
      "Type": "String",
      "Description": "S3 bucket for asset \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    },
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE": {
      "Type": "String",
      "Description": "S3 key for asset version \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    },
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eArtifactHashB9EF04B2": {
      "Type": "String",
      "Description": "Artifact hash for asset \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    }
  }
}